cmd/erasure-server-pool-decom.go

```diff
@@ -0,0 +1,632 @@
+	"github.com/minio/minio/internal/hash"
+	xhttp "github.com/minio/minio/internal/http"
+	"github.com/minio/minio/internal/logger"
+	"github.com/tinylib/msgp/msgp"
+)
+
+// DrainInfo currently draining information
+type DrainInfo struct {
+	StartTime   time.Time `json:"startTime" msg:"st"`
+	StartSize   int64     `json:"startSize" msg:"ssm"`
+	Duration    int64     `json:"duration" msg:"d"`
+	CurrentSize int64     `json:"currentSize" msg:"csm"`
+	Complete    bool      `json:"complete" msg:"cmp"`
+	Failed      bool      `json:"failed" msg:"fl"`
+}
+
+// PoolInfo captures pool info
+type PoolInfo struct {
+	ID         int        `json:"id" msg:"id"`
+	CmdLine    string     `json:"cmdLine" msg:"cli,omitempty"`
+	LastUpdate time.Time  `json:"lastUpdate" msg:"lu"`
+	Drain      *DrainInfo `json:"drainInfo,omitempty" msg:"dr,omitempty"`
+	Suspend    bool       `json:"suspend" msg:"sp"`
+}
+
+//go:generate msgp -file $GOFILE -unexported
+type poolMeta struct {
+	Version string     `msg:"v"`
+	Pools   []PoolInfo `msg:"pls"`
+}
+
+func (p poolMeta) Resume(idx int) bool {
+	if p.IsSuspended(idx) {
+		// We can only resume non-draining pools.
+		if p.Pools[idx].Drain == nil {
+			p.Pools[idx].Suspend = false
+			p.Pools[idx].LastUpdate = time.Now()
+			return true
+		}
+	}
+	return false
+}
+
+func (p poolMeta) DrainComplete(idx int) bool {
+	p.Pools[idx].Suspend = true
+	p.Pools[idx].LastUpdate = time.Now()
+	if p.Pools[idx].Drain != nil {
+		p.Pools[idx].Drain.Complete = true
+		p.Pools[idx].Drain.Failed = false
+		return true
+	}
+	return false
+}
+
+func (p poolMeta) DrainFailed(idx int) bool {
+	p.Pools[idx].Suspend = true
+	p.Pools[idx].LastUpdate = time.Now()
+	if p.Pools[idx].Drain != nil {
+		p.Pools[idx].Drain.Complete = false
+		p.Pools[idx].Drain.Failed = true
+		return true
+	}
+	return false
+}
+
+func (p poolMeta) Drain(idx int, info StorageInfo) bool {
+	if p.Pools[idx].Drain != nil && p.Pools[idx].Drain.Complete {
+		// Draining is complete, do not have to drain again.
+		return false
+	}
+
+	p.Pools[idx].Suspend = true
+	p.Pools[idx].LastUpdate = time.Now()
+	p.Pools[idx].Drain = &DrainInfo{
+		StartTime: time.Now(),
+		StartSize: int64(TotalUsableCapacityFree(info)),
+	}
+	return true
+
+}
+
+func (p poolMeta) Suspend(idx int) bool {
+	if p.IsSuspended(idx) {
+		return false
+	}
+	p.Pools[idx].Suspend = true
+	p.Pools[idx].LastUpdate = time.Now()
+	return true
+}
+
+func (p poolMeta) IsSuspended(idx int) bool {
+	return p.Pools[idx].Suspend
+}
+
+func (p *poolMeta) load(ctx context.Context, set *erasureSets, npools int) (bool, error) {
+	gr, err := set.GetObjectNInfo(ctx, minioMetaBucket, poolMetaName,
+		nil, http.Header{}, readLock, ObjectOptions{})
+	if err != nil && !isErrObjectNotFound(err) {
+		return false, err
+	}
+	if isErrObjectNotFound(err) {
+		return true, nil
+	}
+	defer gr.Close()
+
+	if err = p.DecodeMsg(msgp.NewReader(gr)); err != nil {
+		return false, err
+	}
+
+	switch p.Version {
```

Comment:
Having the version inside the data makes it inflexible, since it is too late to decode into another struct if we should need that at one point.

So I generally prefer to have it before the payload data.

Reply:
Sure okay, let me do xl.meta style.